### PR TITLE
[core] Reworked the CRcvBuffer::dropMessage(..) function.

### PR DESCRIPTION
**Pull request by maxsharabayko**
To improve the `CRcvBuffer::dropMessage(..)` performance the packets are dropped based on pkt seqno range first.
Then, if the start of the message has not been dropped, and additional search step from the `seqnolo` position downwards is applied to look for a packet with `PB_FIRST` boundary flag.

Proceeding after #2657.

@gou4shi1 please have a look.

### CODE REVIEW

**Comment by gou4shi1** [comment]
What if `seqnohi` is not the last packet of the message to drop?
For example, seqno of msg2 is `3-5` but this function is called with `(msgno=2, seqlo=3, seqhi=4)`.
I took a rough look, this case should not happen in the current codebase, but we shuould add some sane checks or at least document this arguments constraint?

**Comment by maxsharabayko** [comment]
> What if `seqnohi` is not the last packet of the message to drop? For example, seqno of msg2 is `3-5` but this function is called with `(msgno=2, seqlo=3, seqhi=4)`. I took a rough look, this case should not happen in the current codebase, but we shuould add some sane checks (for future) or at least document this constraint?

This can only happen in case the sender reports an invalid range. In general, the sender should have at least the last packet of the message it is requesting to be dropped.

**Comment by gou4shi1** [comment]
> In general, the sender should have at least the last packet of the message it is requesting to be dropped.

Agreed, but would be nice to have it documented.

**Review by ethouris — APPROVED** [review]

File: srtcore/buffer_rcv.cpp [hunk 1]
```diff
@@ -257,84 +257,105 @@ int CRcvBuffer::dropAll()
...
+            }
+
+            ++iDropCnt;
+            dropUnitInPos(i);
+            m_entries[i].status = EntryState_Drop;
+
+            if (minDroppedOffset == -1)
+                minDroppedOffset = offPos(m_iStartPos, i);
+            else
+                minDroppedOffset = min(offPos(m_iStartPos, i), minDroppedOffset);
```
[the remark this answers is not in the document]
Comment (maxsharabayko) [reply srtcore/buffer_rcv.cpp:351]: Mmm.. Why? We need to find the minimum offset.

[the remark this answers is not in the document]
Comment (gou4shi1) [reply srtcore/buffer_rcv.cpp:351]: The new offset should always be the smaller one?

[the remark this answers is not in the document]
Comment (maxsharabayko) [reply srtcore/buffer_rcv.cpp:351]: Yes, `minDroppedOffset` is later used to check if the next reading position should be updated.
```c++
const bool needUpdateNonreadPos = (minDroppedOffset != -1 && minDroppedOffset <= getRcvDataSize());
```

[the remark this answers is not in the document]
Comment (gou4shi1) [reply srtcore/buffer_rcv.cpp:351]: I mean the new `offPos(m_iStartPos, i)` should always smaller than `minDroppedOffset`.

[the remark this answers is not in the document]
Comment (maxsharabayko) [reply srtcore/buffer_rcv.cpp:351]: Hm, right. Given the search stays always backward it should be enough to just set `minDroppedOffset = offPos(m_iStartPos, i)`.

**Pull request merged**